### ISSUE TITLE
RFC 7505 Null MX detection is broken

### ISSUE THREAD

**Issue by johannessen** [Perl-Email-Project/Email-Valid#57]
Net::DNS [returns](https://metacpan.org/dist/Net-DNS/source/lib/Net/DNS/Domain.pm#L162) the host name for Null MX records as the string `'.'`. As a result, Null MX records are *never* detected by version 1.203 of Email::Valid.

[valid.t](https://github.com/Perl-Email-Project/Email-Valid/blob/ede466a520a668932bbf4a4f97f6e892ffc39077/t/valid.t#L159-L162) runs a check against `nomail.digilicious.com`, which (currently) has no A record. For the test suite to verify that the Null MX check works correctly, it needs to check a name which has both a Null MX *and* an A record. Otherwise, the missing A record will *always* lead to the address check failing (and the unit test therefore succeeding), with the missing A record masking the result of the Null MX detection.

A DNS name that *currently* has a Null MX and address records would be `ajnn.de`. However, there is no guarantee that it’ll stay that way, so this particular name is at most useful for development testing, not for automated testing.

Since I don’t know which DNS name I should use in the test suite, I haven’t provided a PR to fix this issue. The fix itself is very simple though:

```patch
--- lib/Email/Valid.pm~	2023-05-16 10:40:00.000000000 +0200
+++ lib/Email/Valid.pm	2023-06-29 17:54:00.000000000 +0200
@@ -151,5 +151,5 @@
     # Check for RFC-7505 Null MX
     my $nmx = scalar @mx_entries;
-    if ($nmx == 1 && length($mx_entries[0]->exchange) == 0) {
+    if ($nmx == 1 && $mx_entries[0]->exchange eq '.') {
       return $self->details('mx');
     }
```

https://github.com/Perl-Email-Project/Email-Valid/blob/ede466a520a668932bbf4a4f97f6e892ffc39077/lib/Email/Valid.pm#L151-L155

**Comment by gene-hightower** [comment]
That fix looks right to me.
As for the test requiring a domain with an A record also, that's another reasonable test case.
As is the case with any public DNS (not "mocked" locally for the tests) the results may change.

**Comment by gene-hightower** [comment]
A domain with no A record, but with an MX record that points to another domain that *does* have an A record, should be valid to send mail to, correct?

**Comment by johannessen** [comment]
Yes, that’s correct.

By the way, I just realised that `details()` might allow for the test suite to make a distinction between a failed Null MX detection and a failed A record detection after all.

**Comment by gene-hightower** [comment]
I put an A record on nomail.digilicious.com in case that helps.

**Comment by johannessen** [comment]
At the moment that A record results in tests failing when trying to install Email::Valid, so I’d hold off on that.

**Comment by gene-hightower** [comment]
> At the moment that A record results in tests failing when trying to install Email::Valid, so I’d hold off on that.

Okay, I have removed it.

**Comment by johannessen** [comment]
I think #58 should resolve this.

@gene-hightower Feel free to add the A record back once the change is merged and released on CPAN.

**Comment by rjbs** [comment]
Fixed via modified #58
